### DISCUSSION TITLE
Moddable Two board - GPIO closed?

### DISCUSSION THREAD

**Discussion by rei-vilo** [phoddie/node-red-mcu#49] [Q&A] [answered]
I am testing the LED on GPIO2 of the [Moddable Two board](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/moddable-two.md#pinout) and set GPIO to `2` although it displays then `3`.

<img width="580" alt="Capture 2022-11-29 à 20 24 26" src="https://user-images.githubusercontent.com/951776/204628608-821973c7-f5d2-4af9-beff-04b231622bfe.png">

## Project 1: Blink

<img width="632" alt="Capture 2022-11-29 à 20 28 57" src="https://user-images.githubusercontent.com/951776/204629728-bc0487af-28dd-4030-a8a7-da19fb6d01b9.png">

* Works as expected.
* The GPIO node shows the state `0` or `1`.

## Project 2: On/Off buttons

This is a separate project, no a flow on the same project.

<img width="597" alt="Capture 2022-11-29 à 20 22 55" src="https://user-images.githubusercontent.com/951776/204627840-b9f480cb-5c38-4d1e-be03-d4e9be897a70.png">

* Clicking on `0` or `1` doesn't change the LED.
* The GPIO node shows **Closed**.

What am I doing wrong?

## Listings

<details>
<summary>Listing 1: Blink</summary>

```
[
    {
        "id": "b0430bb22e8d2233",
        "type": "inject",
        "z": "fa3bf03c6582e230",
        "name": "",
        "props": [],
        "repeat": "2",
        "crontab": "",
        "once": true,
        "onceDelay": "1",
        "topic": "",
        "_mcu": {
            "mcu": true
        },
        "x": 130,
        "y": 100,
        "wires": [
            [
                "a3c28be7578a24fc",
                "ea51a1713c293c29"
            ]
        ]
    },
    {
        "id": "a3c28be7578a24fc",
        "type": "delay",
        "z": "fa3bf03c6582e230",
        "name": "",
        "pauseType": "delay",
        "timeout": "500",
        "timeoutUnits": "milliseconds",
        "rate": "1",
        "nbRateUnits": "1",
        "rateUnits": "second",
        "randomFirst": "1",
        "randomLast": "5",
        "randomUnits": "seconds",
        "drop": false,
        "allowrate": false,
        "outputs": 1,
        "_mcu": {
            "mcu": true
        },
        "x": 190,
        "y": 240,
        "wires": [
            [
                "26da7f94dc0f4b7e"
            ]
        ]
    },
    {
        "id": "65c9d70eff75522b",
        "type": "rpi-gpio out",
        "z": "fa3bf03c6582e230",
        "name": "",
        "pin": "2",
        "set": "",
        "level": "0",
        "freq": "",
        "out": "out",
        "bcm": true,
        "_mcu": {
            "mcu": true
        },
        "x": 530,
        "y": 180,
        "wires": []
    },
    {
        "id": "ea51a1713c293c29",
        "type": "change",
        "z": "fa3bf03c6582e230",
        "name": "1",
        "rules": [
            {
                "t": "set",
                "p": "payload",
                "pt": "msg",
                "to": "1",
                "tot": "num"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "_mcu": {
            "mcu": true
        },
        "x": 370,
        "y": 160,
        "wires": [
            [
                "65c9d70eff75522b",
                "9aad4231d9142b20"
            ]
        ]
    },
    {
        "id": "26da7f94dc0f4b7e",
        "type": "change",
        "z": "fa3bf03c6582e230",
        "name": "0",
        "rules": [
            {
                "t": "set",
                "p": "payload",
                "pt": "msg",
                "to": "0",
                "tot": "num"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "_mcu": {
            "mcu": true
        },
        "x": 370,
        "y": 260,
        "wires": [
            [
                "0ae1e6d23a6488ce",
                "65c9d70eff75522b"
            ]
        ]
    },
    {
        "id": "9aad4231d9142b20",
        "type": "debug",
        "z": "fa3bf03c6582e230",
        "name": "debug 1",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "false",
        "statusVal": "",
        "statusType": "auto",
        "_mcu": {
            "mcu": true
        },
        "x": 540,
        "y": 140,
        "wires": []
    },
    {
        "id": "0ae1e6d23a6488ce",
        "type": "debug",
        "z": "fa3bf03c6582e230",
        "name": "debug 2",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "false",
        "statusVal": "",
        "statusType": "auto",
        "_mcu": {
            "mcu": true
        },
        "x": 540,
        "y": 240,
        "wires": []
    }
]
```

</details>

<details>
<summary>Listing 2: On/Off buttons</summary>

```
[
    {
        "id": "af65a212cb5d9857",
        "type": "rpi-gpio out",
        "z": "cc61b7b4ecba6479",
        "name": "",
        "pin": "2",
        "set": true,
        "level": "1",
        "freq": "",
        "out": "out",
        "bcm": true,
        "_mcu": {
            "mcu": true
        },
        "x": 580,
        "y": 220,
        "wires": []
    },
    {
        "id": "2eae9d517eb042a0",
        "type": "inject",
        "z": "cc61b7b4ecba6479",
        "name": "",
        "props": [
            {
                "p": "payload"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "topic": "",
        "payload": "1",
        "payloadType": "num",
        "_mcu": {
            "mcu": true
        },
        "x": 190,
        "y": 120,
        "wires": [
            [
                "af65a212cb5d9857"
            ]
        ]
    },
    {
        "id": "754e7a6b30f2ac26",
        "type": "inject",
        "z": "cc61b7b4ecba6479",
        "name": "",
        "props": [
            {
                "p": "payload"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "topic": "",
        "payload": "0",
        "payloadType": "num",
        "_mcu": {
            "mcu": true
        },
        "x": 190,
        "y": 180,
        "wires": [
            [
                "af65a212cb5d9857"
            ]
        ]
    }
]
```
</details>

**Comment by phoddie** [comment]
Thank you for the very clear report.

When it doesn't work, do you still see the Debug node output traces?

I was able to run both flows on Moddable Two without a problem. To eliminate any possible interaction with the Node-RED Editor, I  modified the flows to use repeating injects to toggle the LED. Maybe you could try that? Here's my modified flow for Listing 2:

<details>

```json
[
    {
        "id": "af65a212cb5d9857",
        "type": "rpi-gpio out",
        "z": "cc61b7b4ecba6479",
        "name": "",
        "pin": "2",
        "set": true,
        "level": "1",
        "freq": "",
        "out": "out",
        "bcm": true,
        "x": 580,
        "y": 220,
        "wires": []
    },
    {
        "id": "2eae9d517eb042a0",
        "type": "inject",
        "z": "cc61b7b4ecba6479",
        "name": "",
        "props": [
            {
                "p": "payload"
            }
        ],
        "repeat": "1",
        "crontab": "",
        "once": true,
        "onceDelay": 0.1,
        "topic": "",
        "payload": "1",
        "payloadType": "num",
        "x": 190,
        "y": 120,
        "wires": [
            [
                "af65a212cb5d9857"
            ]
        ]
    },
    {
        "id": "754e7a6b30f2ac26",
        "type": "inject",
        "z": "cc61b7b4ecba6479",
        "name": "",
        "props": [
            {
                "p": "payload"
            }
        ],
        "repeat": "1",
        "crontab": "",
        "once": true,
        "onceDelay": 0.6,
        "topic": "",
        "payload": "0",
        "payloadType": "num",
        "x": 190,
        "y": 180,
        "wires": [
            [
                "af65a212cb5d9857"
            ]
        ]
    }
]
```

</details>


I have no idea why the Editor shows the node as "PIN: 3". That's really strange.

**Reply by ralphwetzel** [reply]
> I have no idea why the Editor shows the node as "PIN: 3". That's really strange.

<img width="366" alt="image" src="https://user-images.githubusercontent.com/16342003/204646206-dc0e7c37-3c43-4a2c-9219-bdc4efacbec1.png">

GPIO 2 == @ PIN 3

**Reply by phoddie** [reply]
Oh, I see. It is a Raspberry Pi thing. That makes sense. I like using the pre-existing Node, because it makes projects more portable and gives developers less to learn. But... details like this are kind of confusing too. Maybe in time we can evolve the editor experience to adapt better to the target device.

**Reply by ralphwetzel** [reply]
Once I drafted a "generic" GPIO node dedicated to `node-red-mcu` - even with that nice looking RPi like interface. Maybe in time ... 😄

**Comment by rei-vilo** [comment, marked as answer]
After much investigation, I managed to get **Project 2: On/Off buttons** working.

The hint was that most of the errors I have reported so far seem to be random, the worst case of errors. 

The trick was: 

* I deleted the `/Users/USER/Projects/moddable/build/tmp/esp32/moddable_two/debug/` folder to get a clean environment.  

This is actually the same fix I used for https://github.com/phoddie/node-red-mcu/discussions/46#discussioncomment-4257754.

It would be a good idea to include that step into the standard build procedure.

**Reply by phoddie** [reply]
Glad you solved this. You should only need to remove the `tmp` build results if you updated the Moddable SDK. That's necessary when the tools have changed. That is step 2 in the Updating [documentation](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/Moddable%20SDK%20-%20Getting%20Started.md#updating).

That said, it is not uncommon to forget to do that. We've made some changes to the Moddable SDK to report an  error when the binary tools are an older build than the current SDK. Those will roll out shortly, and should help to catch problems like this.

**Reply by rei-vilo** [reply]
Yes, indeed :)

Dealing with parallel procedures to install the Moddable SDK,  the tool-chains and SDKs for each target (here, Raspberry Pi Pico and ESP32), is not easy.

**Reply by phoddie** [reply]
> Dealing with parallel procedures to install the Moddable SDK, the tool-chains and SDKs for each target (here, Raspberry Pi Pico and ESP32), is not easy

Understood. That's a reality of embedded development in general. Each silicon vendor has entirely different tools. We are doing our best in the Moddable SDK to smooth that out, but we cannot make the problem entirely evaporate. I'm optimistic that in time [xs-dev](https://github.com/hipsterbrown/xs-dev) will automate this for the majority of developers.

**Reply by rei-vilo** [reply]
No problem, I am well aware of the highly segmented world of embedded development. 

However, could a container with the tools for the official boards sold by Moddable be a solution? 

I mean, with limited resources, each minute spent on configuring the environment is not dedicated to evaluate the Moddable SDK.

**Reply by phoddie** [reply]
> However, could a container with the tools for the official boards sold by Moddable be a solution?

A container could support a lot of boards, just as easily as the Moddable boards. We looked at providing a container a while back, it wasn't obvious what would work for most developers. We are now doing monthly releases that include binaries. Maybe we should revisit providing a container as part of that. The Moddable team doesn't have much experience with containers, so we probably aren't the right people to take the lead on it though. Is that something you might be able to help set-up? From there, we could update it to include as part of our monthly release.

FWIW – I recall that @meganetaaan had, at one point, experimented with providing a container for his (super-~fun~kawaii, award-winning) Stack-chan project. I'm not sure about the status of that now.

**Reply by rei-vilo** [reply]
About containers, I am mostly a user. There are two main benefits of containers: easy installation and separate environment.

The recommended [xs-dev](https://github.com/HipsterBrown/xs-dev) provide a good solution for installing the SDKs, albeit not in a separate environment.

**Reply by phoddie** [reply]
Cool. I'm glad to hear that about xs-dev. I understand it isn't the same as a container but it does automate installs and updates. xs-dev might even be a good tool for someone to use to build a container.

**Reply by rei-vilo** [reply]
I have just performed a clean installation with `xs-dev`. 

It is worth noting that `xs-dev` only manage the Moddable and the boards-related SDKs. 

The Node-RED MCU plug-in and the `settings.js` file editing need to be performed manually.

**Comment by rei-vilo** [comment]
One more question related to GPIOs, this time `PWM 18` in charge of [backlight](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/moddable-two.md#backlight) from Node-RED MCU.

+ Configuration the RPI-GPIO node with `PWM 18` shows `12 not available`. 
<img width="153" alt="Capture 2022-12-14 à 18 18 19" src="https://user-images.githubusercontent.com/951776/207663239-94b23fb8-214a-483b-8830-23f9c11327cb.png">

+ Trying to access `config.brightness = 75;" or `backlight.write(80);` return errors as well.
<img width="151" alt="Capture 2022-12-14 à 18 18 47" src="https://user-images.githubusercontent.com/951776/207663341-1e6c3b45-b122-40e8-826d-2a506ff1fed0.png">

+ The panes **Configuration nodes** and **Context** do not list this parameter or are empty.

How to manage backlight with Node-RED?

**Reply by phoddie** [reply]
I think there's a straightforward solution here, but won't be able to look into it for a bit. Since it is a different topic, could you move this question to a separate topic? That way it will be easier for others to discover the knowledge in the future. Thanks!

**Reply by rei-vilo** [reply]
Sure, moved to #56.
